File: src/desktop/legacy/placement_managers/attr_key_generator.py
from data.constants import CLOCK, COUNTER, IN, OUT
from enums.letter.letter import Letter, LetterCondition
from placement_managers.arrow_placement_manager.arrow_placement_context import (
    ArrowPlacementContext,
)
from objects.arrow.arrow import Arrow


class AttrKeyGenerator:
    def __init__(self):
        pass  # No dependency on `SpecialArrowPositioner` needed anymore

    def get_key_from_arrow(self, arrow: "Arrow") -> str:
        """Original method for getting key from Arrow (kept for compatibility)."""
        return self.generate_key(
            motion_type=arrow.motion.state.motion_type,
            letter=arrow.pictograph.state.letter,
            start_ori=arrow.motion.state.start_ori,
            color=arrow.state.color,
            lead_state=arrow.motion.state.lead_state,
            has_hybrid_motions=arrow.pictograph.managers.check.has_hybrid_motions(),
            starts_from_mixed_orientation=arrow.pictograph.managers.check.starts_from_mixed_orientation(),
            starts_from_standard_orientation=arrow.pictograph.managers.check.starts_from_standard_orientation(),
        )

    def get_key_from_context(self, context: ArrowPlacementContext) -> str:
        """New method to get key from ArrowPlacementContext instead of Arrow object."""
        return self.generate_key(
            motion_type=context.motion_type,
            letter=context.letter,
            start_ori=None,  # No equivalent in ArrowPlacementContext, assume None
            color=context.arrow_color,
            lead_state=None,  # Not available in context, assume None
            has_hybrid_motions=False,  # Need a better check here if hybrid motions are relevant
            starts_from_mixed_orientation=False,  # Context lacks a direct check for this
            starts_from_standard_orientation=True,  # Assume standard unless explicitly hybrid
        )

    def generate_key(
        self,
        motion_type: str,
        letter: Letter,
        start_ori: str,
        color: str,
        lead_state: str,
        has_hybrid_motions: bool,
        starts_from_mixed_orientation: bool,
        starts_from_standard_orientation: bool,
    ) -> str:
        """Core key generation logic used by both `get_key` and `get_key_from_context`."""
        if starts_from_mixed_orientation:
            if letter in ["S", "T"]:
                return f"{lead_state}"
            elif has_hybrid_motions:
                if start_ori in [IN, OUT]:
                    return f"{motion_type}_from_layer1"
                elif start_ori in [CLOCK, COUNTER]:
                    return f"{motion_type}_from_layer2"
                else:
                    return color
            elif letter in letter.get_letters_by_condition(LetterCondition.NON_HYBRID):
                return color
            else:
                return motion_type

        elif starts_from_standard_orientation:
            if letter in ["S", "T"]:
                return f"{color}_{lead_state}"
            elif has_hybrid_motions:
                return motion_type
            else:
                return color

        return motion_type  # Default fallback

File: legacy_web_app/src/lib/components/objects/Arrow/ArrowPlacementManager/utils/AttrKeyGenerator.ts
import type { ArrowData } from '$lib/components/objects/Arrow/ArrowData';
import type { PictographData } from '$lib/types/PictographData';
import { LetterUtils } from '$lib/utils/LetterUtils';
import { LetterConditions } from '$lib/components/Pictograph/constants/LetterConditions';
import { CLOCK, COUNTER, IN, OUT } from '$lib/types/Constants';
import type { PictographChecker } from '$lib/components/Pictograph/services/PictographChecker';

export class AttrKeyGenerator {
	private static instance: AttrKeyGenerator;

	private constructor() {}

	public static getInstance(): AttrKeyGenerator {
		if (!this.instance) {
			this.instance = new AttrKeyGenerator();
		}
		return this.instance;
	}

	/**
	 * Generates a key for an arrow based on multiple conditions
	 * @param arrow The arrow data
	 * @param checker The PictographChecker instance
	 * @param pictographData The full pictograph data
	 * @returns The generated key string
	 */
	public getKeyFromArrow(
		arrow: ArrowData,
		checker: PictographChecker,
		pictographData: PictographData
	): string {
		const letter = pictographData.letter;
		if (!letter) return arrow.color;

		const letterEnum = LetterUtils.fromString(letter);
		const hasHybridMotions = this.hasHybridMotions(pictographData);

		// Use helper method to determine starts from orientation
		const startsFromMixedOrientation = this.startsFromMixedOrientation(pictographData);
		const startsFromStandardOrientation = !startsFromMixedOrientation;

		const startOri = this.getStartOrientation(arrow, pictographData);
		const leadState = this.getLeadState(pictographData);

		// Key generation logic matching Python implementation
		if (startsFromMixedOrientation) {
			if (['S', 'T'].includes(letter)) {
				return leadState || '';
			}

			if (hasHybridMotions) {
				if ([IN, OUT].includes(startOri)) {
					return `${arrow.motionType}_from_layer1`;
				} else if ([CLOCK, COUNTER].includes(startOri)) {
					return `${arrow.motionType}_from_layer2`;
				} else {
					return arrow.color;
				}
			}

			if (LetterUtils.getLettersByCondition(LetterConditions.NON_HYBRID).includes(letterEnum)) {
				return arrow.color;
			}

			return arrow.motionType;
		}

		if (startsFromStandardOrientation) {
			if (['S', 'T'].includes(letter)) {
				return `${arrow.color}${leadState}`;
			}

			if (hasHybridMotions) {
				return arrow.motionType;
			}

			return arrow.color;
		}

		return arrow.motionType; // Default fallback
	}

	private getStartOrientation(arrow: ArrowData, pictographData: PictographData): string {
		// Determine start orientation based on motion color
		const motionData =
			arrow.color === 'red' ? pictographData.redMotionData : pictographData.blueMotionData;

		return motionData?.startOri || '';
	}

	private getLeadState(pictographData: PictographData): string | null {
		// Determine lead state, prioritizing float motions
		const redMotion = pictographData.redMotionData;
		const blueMotion = pictographData.blueMotionData;

		if (redMotion?.motionType === 'float') return 'red_float';
		if (blueMotion?.motionType === 'float') return 'blue_float';

		return null;
	}

	private hasHybridMotions(pictographData: PictographData): boolean {
		// Check if pictograph has hybrid motions
		const letter = pictographData.letter;
		if (!letter) return false;

		const letterEnum = LetterUtils.fromString(letter);
		return LetterUtils.getLettersByCondition(LetterConditions.HYBRID).includes(letterEnum);
	}

	private startsFromMixedOrientation(pictographData: PictographData): boolean {
		const redMotion = pictographData.redMotionData;
		const blueMotion = pictographData.blueMotionData;

		if (!redMotion || !blueMotion) return false;

		const redStartOri = redMotion.startOri;
		const blueStartOri = blueMotion.startOri;

		// Check for different orientation types
		const radialOrientations = [IN, OUT];
		const nonradialOrientations = [CLOCK, COUNTER];

		const redIsRadial = radialOrientations.includes(redStartOri);
		const redIsNonRadial = nonradialOrientations.includes(redStartOri);
		const blueIsRadial = radialOrientations.includes(blueStartOri);
		const blueIsNonRadial = nonradialOrientations.includes(blueStartOri);

		return (redIsRadial && blueIsNonRadial) || (redIsNonRadial && blueIsRadial);
	}
}
